src/sdam/server_selection.ts

```diff
@@ -28,6 +31,24 @@ export function writableServerSelector(): ServerSelector {
     );
 }
 
+/**
+ * Returns a server selector that uses a read preference to select a
+ * server potentially for a write on a secondary.
+ */
+export function secondaryWritableServerSelector(
```

Comment:
Our new custom server selector that determines the effective read preference for server selection and delegates to the existing `readPreferenceServerSelector`. Implementing a new one here and delegating allows us not to mess with the existing logic in the other selector. While it is backed by spec tests, the lack of unit tests was a red flag for me to approach it by modifying the other.